feat(website): new documentation (#1048)
algolia/docsearch: packages/website/docs/record-extractor.md

---
title: Record Extractor
---

:::info

The following content is for **[DocSearch v3][2]** and **[its new infrastructure][5]**. If you are using **[DocSearch v2][3]** or the **[docsearch-scraper][6]**, see the **[legacy][4]** documentation.

:::

## Introduction

:::info

> This documentation will only contain informations regarding the **helpers.docsearch** method, see **[Algolia Crawler Documentation][7]** for more informations on the **[Algolia Crawler][8]**.

Reviewer: ```suggestion
This documentation will only contain information regarding the **helpers.docsearch** method, see **[Algolia Crawler Documentation][7]** for more information on the **[Algolia Crawler][8]**.
```

### Useful links

- [Extracting records with the Algolia Crawler][11]
- [`recordExtractor` parameters][10]

## Usage

The most common way to use the DocSearch helper, is to return its result to the [`recordExtractor`][9] function.

```js
recordExtractor: ({ helpers }) => {
  return helpers.docsearch({
    recordProps: {
      lvl0: {
        selectors: "header h1",
      },
      lvl1: "article h2",
      lvl2: "article h3",
      lvl3: "article h4",
      lvl4: "article h5",
      lvl5: "article h6",
      content: "main p, main li",
    },
  });
},
```

## Complex extractors

### Using the Cheerio instance (`$`)

You can also use the provided [`Cheerio instance ($)`][14] to exclude content from the DOM:

```js
recordExtractor: ({ $, helpers }) => {
  // Removing DOM elements we don't want to crawl
  $(".my-warning-message").remove();

  return helpers.docsearch({
    recordProps: {
      lvl0: {
        selectors: "header h1",
      },
      lvl1: "article h2",
      lvl2: "article h3",
      lvl3: "article h4",
      lvl4: "article h5",
      lvl5: "article h6",
      content: "main p, main li",
    },
  });
},
```

### With fallback DOM selectors

> Each `lvlX` and `content` supports fallback selectors as an array of string, which allow robust config files:

Reviewer: ```suggestion
Each `lvlX` and `content` supports fallback selectors as an array of string, which allows for robust config files:
```

```js
recordExtractor: ({ $, helpers }) => {
  return helpers.docsearch({
    recordProps: {
      // `.exists h1` will be selected if `.exists-probably h1` does not exists.
      lvl0: {
        selectors: [".exists-probably h1", ".exists h1"],
      }
      lvl1: "article h2",
      lvl2: "article h3",
      lvl3: "article h4",
      lvl4: "article h5",
      lvl5: "article h6",
      // `.exists p, .exists li` will be selected.
      content: [
        ".does-not-exists p, .does-not-exists li",
        ".exists p, .exists li",
      ],
    },
  });
},
```

### With custom variables

Custom variables are useful to filter content in the frontend (`version`, `lang`, etc.).

> _These selectors also supports [`defaultValue`](#with-raw-text-defaultvalue) and [fallback selectors](#with-fallback-dom-selectors)_

Reviewer: ```suggestion
_These selectors also support [`defaultValue`](#with-raw-text-defaultvalue) and [fallback selectors](#with-fallback-dom-selectors)_
```
